cmd/kubectl/kubectlcobra/prune.go

```diff
@@ -0,0 +1,246 @@
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/printers"
+	"k8s.io/cli-runtime/pkg/resource"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/kubectl/pkg/cmd/apply"
+	"k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/validation"
+)
+
+// PruneOptions encapsulates the necessary information to
+// implement the prune functionality.
+type PruneOptions struct {
+	Client    dynamic.Interface
+	Builder   *resource.Builder
+	Mapper    meta.RESTMapper
+	Namespace string
+	// The currently applied objects (as Infos), including the
+	// current grouping object. These objects are used to
+	// calculate the prune set after retreiving the previous
+	// grouping objects.
+	CurrentGroupingObject *resource.Info
+	// The set of retrieved grouping objects (as Infos) selected
+	// by the grouping label. This set should also include the
+	// current grouping object. Stored here to make testing
+	// easier by manually setting the retrieved grouping infos.
+	PastGroupingObjects      []*resource.Info
+	RetrievedGroupingObjects bool
+
+	ToPrinter func(string) (printers.ResourcePrinter, error)
+	Out       io.Writer
+
+	Validator validation.Schema
+
+	// TODO: DeleteOptions--cascade?
+}
+
+// NewPruneOptions returns a struct (PruneOptions) encapsulating the necessary
+// information to run the prune. Returns an error if an error occurs
+// gathering this information.
+// TODO: Add dry-run options.
+func NewPruneOptions(f util.Factory, o *apply.ApplyOptions) (*PruneOptions, error) {
+
+	po := &PruneOptions{}
+	var err error
+
+	po.Client, err = f.DynamicClient()
+	if err != nil {
+		return nil, err
+	}
+	po.Builder = f.NewBuilder()
+	po.Mapper, err = f.ToRESTMapper()
+	if err != nil {
+		return nil, err
+	}
+	po.Namespace = o.Namespace
+	currentObjects, err := o.GetObjects()
+	if err != nil {
+		return nil, err
+	}
+	currentGroupingObject, found := findGroupingObject(currentObjects)
+	if !found {
+		return nil, fmt.Errorf("Current grouping object not found during prune.")
+	}
+	po.CurrentGroupingObject = currentGroupingObject
+	po.PastGroupingObjects = []*resource.Info{}
+	po.RetrievedGroupingObjects = false
+
+	po.ToPrinter = o.ToPrinter
+	po.Out = o.Out
+
+	po.Validator, err = f.Validator(false)
+	if err != nil {
+		return nil, err
+	}
+
+	return po, nil
+}
+
+// GetPreviousGroupingObjects returns the set of grouping objects
+// that have the same label as the current grouping object. Removes
+// the current grouping object from this set. Returns an error
+// if there is a problem retrieving the grouping objects.
+func (po *PruneOptions) GetPreviousGroupingObjects() ([]*resource.Info, error) {
+
+	// Get the grouping label for this grouping object, and create
+	// a label selector from it.
+	groupingLabel, err := retrieveGroupingLabel(po.CurrentGroupingObject.Object)
+	if err != nil {
+		return nil, err
+	}
+	labelSelector := fmt.Sprintf("%s=%s", GroupingLabel, groupingLabel)
```

Comment:
The above code should be inside the if block below (afaict, it's not used otherwise).